libr/debug/p/native/windows/windows_debug.c

```diff
@@ -0,0 +1,1312 @@
+	if (!path) {
+		path = r_sys_conv_win_to_utf8 (name);
+	}
+	// it is possible to get pc for a non debugged process but the operation is expensive and might be risky
+	RDebugPid *ret = r_debug_pid_new (path, pid, uid, 's', 0);
+	free (path);
+	return ret;
+}
+
+RList *w32_pid_list(RDebug *dbg, int pid, RList *list) {
+	HANDLE sh = CreateToolhelp32Snapshot (TH32CS_SNAPPROCESS, pid);
+	if (sh == INVALID_HANDLE_VALUE) {
+		r_sys_perror ("w32_pid_list/CreateToolhelp32Snapshot");
+		return list;
+	}
+	PROCESSENTRY32 pe;
+	pe.dwSize = sizeof (pe);
+	if (Process32First (sh, &pe)) {
+		RIOW32Dbg *rio = dbg->user;
+		bool all = pid == 0, b = false;
+		do {
+			if (all || pe.th32ProcessID == pid || (b = pe.th32ParentProcessID == pid)) {
+				// Returns NULL if process is inaccessible unless if its a child process of debugged process
+				RDebugPid *dbg_pid = __build_debug_pid (pe.th32ProcessID, b ? rio->ph : NULL, pe.szExeFile);
+				if (dbg_pid) {
+					r_list_append (list, dbg_pid);
+				}
+#if 0
+				else {
+					eprintf ("w32_pid_list: failed to process pid %d\n", pe.th32ProcessID);
+				}
+#endif
+			}
+		} while (Process32Next (sh, &pe));
+	} else {
+		r_sys_perror ("w32_pid_list/Process32First");
+	}
+	CloseHandle (sh);
+	return list;
+}
+
+RList *w32_desc_list(int pid) {
+	RDebugDesc *desc;
+	RList *ret = r_list_new ();
+	int i;
+	HANDLE ph;
+	PSYSTEM_HANDLE_INFORMATION handleInfo;
+	NTSTATUS status;
+	ULONG handleInfoSize = 0x10000;
+	LPVOID buff;
+	if (!ret) {
+		perror ("win_desc_list/r_list_new");
+		return NULL;
+	}
+	if (!(ph = OpenProcess (PROCESS_DUP_HANDLE, FALSE, pid))) {
+		r_sys_perror ("win_desc_list/OpenProcess");
+		r_list_free (ret);
+		return NULL;
+	}
+	handleInfo = (PSYSTEM_HANDLE_INFORMATION)malloc (handleInfoSize);
+	#define STATUS_INFO_LENGTH_MISMATCH 0xc0000004
+	#define SystemHandleInformation 16
+	while ((status = w32_NtQuerySystemInformation (SystemHandleInformation, handleInfo, handleInfoSize, NULL)) == STATUS_INFO_LENGTH_MISMATCH)
+		handleInfo = (PSYSTEM_HANDLE_INFORMATION)realloc (handleInfo, handleInfoSize *= 2);
+	if (status) {
+		r_sys_perror ("win_desc_list/NtQuerySystemInformation");
+		CloseHandle (ph);
+		r_list_free (ret);
+		return NULL;
+	}
+	for (i = 0; i < handleInfo->HandleCount; i++) {
+		SYSTEM_HANDLE handle = handleInfo->Handles[i];
+		HANDLE dupHandle = NULL;
+		POBJECT_TYPE_INFORMATION objectTypeInfo;
+		PVOID objectNameInfo;
+		UNICODE_STRING objectName;
+		ULONG returnLength;
+		if (handle.ProcessId != pid) {
+			continue;
+		}
+		if (handle.ObjectTypeNumber != 0x1c) {
+			continue;
+		}
+		if (w32_NtDuplicateObject (ph, &handle.Handle, GetCurrentProcess (), &dupHandle, 0, 0, 0)) {
+			continue;
+		}
+		objectTypeInfo = (POBJECT_TYPE_INFORMATION)malloc (0x1000);
+		if (w32_NtQueryObject (dupHandle, 2, objectTypeInfo, 0x1000, NULL)) {
+			CloseHandle (dupHandle);
+			continue;
+		}
+		objectNameInfo = malloc (0x1000);
+		if (w32_NtQueryObject (dupHandle, 1, objectNameInfo, 0x1000, &returnLength)) {
```

Comment:
This function can hang indefinitely on certain types of handles. See the workaround here (create a new thread to query the name and terminate if it doesnt return in a fix amount of time): https://github.com/x64dbg/x64dbg/blob/development/src/dbg/handles.cpp#L58